### PR TITLE
r/aws_connect_phone_number

### PR DESCRIPTION
**Pull request by GlennChia**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #26015

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccConnect_serial/PhoneNumber' PKG=connect
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/connect/... -v -count 1 -parallel 20  -run=TestAccConnect_serial/PhoneNumber -timeout 180m
=== RUN   TestAccConnect_serial
=== RUN   TestAccConnect_serial/PhoneNumber
=== RUN   TestAccConnect_serial/PhoneNumber/prefix
=== RUN   TestAccConnect_serial/PhoneNumber/targetARN
=== RUN   TestAccConnect_serial/PhoneNumber/basic
=== RUN   TestAccConnect_serial/PhoneNumber/disappears
=== RUN   TestAccConnect_serial/PhoneNumber/tags
=== RUN   TestAccConnect_serial/PhoneNumber/description
--- PASS: TestAccConnect_serial (616.97s)
    --- PASS: TestAccConnect_serial/PhoneNumber (616.97s)
        --- PASS: TestAccConnect_serial/PhoneNumber/prefix (97.80s)
        --- PASS: TestAccConnect_serial/PhoneNumber/targetARN (116.53s)
        --- PASS: TestAccConnect_serial/PhoneNumber/basic (94.70s)
        --- PASS: TestAccConnect_serial/PhoneNumber/disappears (90.76s)
        --- PASS: TestAccConnect_serial/PhoneNumber/tags (122.56s)
        --- PASS: TestAccConnect_serial/PhoneNumber/description (94.62s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/connect    617.031s

...
```

### CODE REVIEW

File: internal/service/connect/phone_number.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+	phoneNumberId := d.Id()
+
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("generating uuid for ClientToken for Phone Number %s: %s", phoneNumberId, err)
+	}
+
+	_, err = conn.ReleasePhoneNumberWithContext(ctx, &connect.ReleasePhoneNumberInput{
+		ClientToken:   aws.String(uuid),
```
Comment (schuylr) [inline internal/service/connect/phone_number.go:266]:
```suggestion
		ClientToken:   resource.UniqueId(),
```

File: internal/service/connect/phone_number.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+	phoneNumberId := d.Id()
+
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("generating uuid for ClientToken for Phone Number %s: %s", phoneNumberId, err)
+	}
+
+	if d.HasChange("target_arn") {
+		_, err := conn.UpdatePhoneNumberWithContext(ctx, &connect.UpdatePhoneNumberInput{
+			ClientToken:   aws.String(uuid),
```
Comment (schuylr) [inline internal/service/connect/phone_number.go:230]:
```suggestion
			ClientToken:   resource.UniqueId(),
```

File: internal/service/connect/phone_number.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+	phoneNumber := output.AvailableNumbersList[0].PhoneNumber
+
+	uuid, err := uuid.GenerateUUID()
+	if err != nil {
+		return diag.Errorf("generating uuid for ClientToken for Connect Instance (%s,%s): %s", targetArn, aws.StringValue(phoneNumber), err)
+	}
+
+	input2 := &connect.ClaimPhoneNumberInput{
+		ClientToken: aws.String(uuid), // can't use aws.String(resource.UniqueId()), because not a valid uuid
```
Comment (schuylr) [inline internal/service/connect/phone_number.go:132]: I don't believe there's a required UUID format for ClientToken: https://docs.aws.amazon.com/connect/latest/APIReference/API_ClaimPhoneNumber.html#API_ClaimPhoneNumber_RequestSyntax 

As long as it's less than 500 characters.

**Review by schuylr — CHANGES_REQUESTED** [review]
Thanks for the PR! I think this is a great first pass, but there are some levels of control that are being restricted from the Terraform operator.

I believe the core thing to allow would be the option to provide a full `phone_number` per the [ClaimPhoneNumber](https://docs.aws.amazon.com/connect/latest/APIReference/API_ClaimPhoneNumber.html) API.

More than happy to contribute to this with access provided to your branch.

File: internal/service/connect/phone_number.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+		return diag.Errorf("searching Connect Phone Number for Connect Instance (%s,%s): %s", targetArn, phoneNumberType, err)
+	}
+
+	if output == nil || output.AvailableNumbersList == nil || len(output.AvailableNumbersList) == 0 {
+		return diag.Errorf("searching Connect Phone Number for Connect Instance (%s,%s): empty output", targetArn, phoneNumberType)
+	}
+
+	phoneNumber := output.AvailableNumbersList[0].PhoneNumber
+
+	uuid, err := uuid.GenerateUUID()
```
Comment (schuylr) [inline internal/service/connect/phone_number.go:126]: Suggest this be turned into a common function for the service

**Review by schuylr — APPROVED** [review]
Discussed offline with @GlennChia about proceeding with this functionality for now. Will create another fast-follow issue to allow support for direct `phone_number` supplying.

Will also need to create two data sources which also depend on this PR. Will link issues as pending from this PR as well!

Thanks for the awesome discussion!

File: internal/service/connect/phone_number.go (hunk 4 above)
Comment (schuylr) [reply internal/service/connect/phone_number.go:126]: Agreed offline that the `err` needs to be handled within the function as a return out.

File: internal/service/connect/phone_number.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+		_, err := conn.UpdatePhoneNumberWithContext(ctx, &connect.UpdatePhoneNumberInput{
+			ClientToken:   aws.String(uuid),
+			PhoneNumberId: aws.String(phoneNumberId),
+			TargetArn:     aws.String(d.Get("target_arn").(string)),
+		})
+
+		if err != nil {
+			return diag.Errorf("updating Phone Number (%s): %s", d.Id(), err)
+		}
+
```
Comment (schuylr) [inline internal/service/connect/phone_number.go:238]:
```suggestion
```

**Review by schuylr — APPROVED** [review]
Quick linter fix suggestion

**Comment by ewbankkit** [comment]
@GlennChia Thanks for the contribution :tada: :clap:.

**Pull request merged**